### PR TITLE
Plans: Refactor cart plan upsells to use pricing from data-stores

### PR DESCRIPTION
**Pull request by chriskmnds**
Part of addressing https://github.com/Automattic/wp-calypso/issues/86638

## Proposed Changes

* Refactors `CartFreeUserPlanUpsell` to function form, along with some cleanup
* Imports pricing hook from the Plans data-store (`usePricingMetaForGridPlans`)
* Removes the `get-plan-price` selector as no longer used

## Media

Screenshot 2024-04-11 at 2 22 04 PM

## Testing Instructions

There are a couple of messages being rendered through this upsell.

- with a site on a free plan, go to `/domains/[ site ]` and find a domain to purchase
- proceed to checkout without a plan in the cart
- confirm that the plan upsell renders as per media above
- confirm the same with a currency that results in a discount e.g. INR (just set it through SA and check back - no need to repeat or reload)

## Pre-merge Checklist

- [ ] Has the general commit checklist been followed? (PCYsg-hS-p2)
- [ ] https://wpcalypso.wordpress.com/devdocs/docs/testing/index.md for your changes?
- [ ] Have you tested the feature in Simple (P9HQHe-k8-p2), Atomic (P9HQHe-jW-p2), and self-hosted Jetpack sites (PCYsg-g6b-p2)?
- [ ] Have you checked for TypeScript, React or other console errors?
- [ ] Have you used memoizing on expensive computations? More info in [Memoizing with create-selector](https://github.com/Automattic/wp-calypso/blob/trunk/packages/state-utils/src/create-selector/README.md) and [Using memoizing selectors](https://react-redux.js.org/api/hooks#using-memoizing-selectors) and [Our Approach to Data](https://github.com/Automattic/wp-calypso/blob/trunk/docs/our-approach-to-data.md)
- [ ] Have we added the "[Status] String Freeze" label as soon as any new strings were ready for translation (p4TIVU-5Jq-p2)?
- [ ] For changes affecting Jetpack: Have we added the "[Status] Needs Privacy Updates" label if this pull request changes what data or activity we track or use (p4TIVU-aUh-p2)?

### CODE REVIEW

File: client/my-sites/checkout/src/components/test/secondary-cart-promotions.tsx
Comment (chriskmnds) [inline client/my-sites/checkout/src/components/test/secondary-cart-promotions.tsx]: @sirbrillig you may not like the way I mock the data here. We have a couple of options I think - we could import the mocks from `/packages/data-stores/src/plans/mocks/next`, or mock the pricing hook `usePricingMetaForGridPlans`, or mock the underlying endpoints? Not sure what's preferred. WDYT?

File: client/my-sites/checkout/src/components/test/secondary-cart-promotions.tsx
Comment (chriskmnds) [inline client/my-sites/checkout/src/components/test/secondary-cart-promotions.tsx:50]: @sirbrillig previous comment outdated already. 😬 
Some updates in the last commit -> now using and mocking only the pricing-meta hook.

File: client/my-sites/checkout/src/components/test/secondary-cart-promotions.tsx
Comment (chriskmnds) [reply client/my-sites/checkout/src/components/test/secondary-cart-promotions.tsx]: https://github.com/Automattic/wp-calypso/pull/89472#discussion_r1562357889

File: client/my-sites/checkout/cart/cart-free-user-plan-upsell.tsx
Comment (sirbrillig) [inline client/my-sites/checkout/cart/cart-free-user-plan-upsell.tsx:55]: Super minor but based on the types you don't need to use optional chaining here. A cart object is a required prop and it always has a `products` array.

```suggestion
	const firstDomain = cart.products.find( isDomainRegistrationOrTransfer );
```

File: client/my-sites/checkout/cart/cart-free-user-plan-upsell.tsx
Comment (sirbrillig) [inline client/my-sites/checkout/cart/cart-free-user-plan-upsell.tsx:74]: Minor but you don't need `else` here because the previous block has a `return`.

```suggestion
	}
	if ( planPrice && firstDomain && planPrice < firstDomain.item_subtotal_integer ) {
```

File: client/my-sites/checkout/src/components/test/secondary-cart-promotions.tsx
Comment (sirbrillig) [inline client/my-sites/checkout/src/components/test/secondary-cart-promotions.tsx:38]: I've tried this before and run into occasional problems. If you only create the `QueryClient` once, it will persist across tests which can cause data leak and weird test failures or worse, false successes. It's better to recreate it for every test. I suggest putting its definition inside the Wrapper memoized; something like this:

https://github.com/Automattic/wp-calypso/blob/8a937aeb19c6d55d72de361df706b9019f8fdf81/client/my-sites/checkout/src/test/existing-credit-card.tsx#L22

File: client/my-sites/checkout/src/components/test/secondary-cart-promotions.tsx
Comment (sirbrillig) [reply client/my-sites/checkout/src/components/test/secondary-cart-promotions.tsx:50]: I think this should be ok. Thanks for asking about it.

**Review by sirbrillig — COMMENTED** [review]
I didn't test this (I leave that to your team) but from a Payments perspective this looks good. Always nice to see fewer class components and Redux selectors! I left some non-blocking comments but also one that you should definitely do to make sure the tests don't share data.

**Review by aneeshd16 — CHANGES_REQUESTED** [review]
I added a domain for a free site, and I am seeing differences between this PR and production when selecting a longer term for the domain:

Production:
image

This PR:
image

For reference, here's the cart with the Starter plan added:
image

The calculation done in this PR is incorrect since it uses the total cost of the domain product instead of just the first-year cost to calculate the price difference. However, in this case, the calculation done in production is also incorrect since it considers the discounted price of the dot blog domain (although that is a separate issue outside the scope of this PR).

Trying with a 4-year dotcom domain, which costs the same as a Starter plan:

Production:
image

This PR:
image

**Comment by chriskmnds** [comment]
I'm closing this. This is too far into conflict land right now to resurface - given https://github.com/Automattic/wp-calypso/pull/89874 did the same level of component refactors. I will leave it to @sirbrillig to cherry-pick any relevant bits to remove the Redux state from there (basically the remaining parts done in this PR). 

I guess we could have nudged for this PR to be pushed forward if fixes were needed to the underlying logic, not redo the same refactor in another, completely creapling the work here 😞

**Comment by chriskmnds** [comment]
I'm closing this. This is too far into conflict land right now to resurface - given https://github.com/Automattic/wp-calypso/pull/89874 did the same level of component refactors. I will leave it to @sirbrillig to cherry-pick any relevant bits to remove the Redux state from there (basically the remaining parts done in this PR). 

I guess we could have nudged for this PR to be pushed forward if fixes were needed to the underlying logic, not redo the same refactor in another, completely creapling the work here 😞

**Comment by sirbrillig** [comment]
I can't believe that I made nearly the same PR as this, totally from scratch, a week after I had reviewed this one. 😱 I apologize greatly for that. I should have looped your team in on the bug fix for the upsell.

As for the bug that @aneeshd16 mentioned above, I believe https://github.com/Automattic/wp-calypso/pull/90893 should fix it, but I'll wait for your review to be sure.

As for the remaining differences between this PR and https://github.com/Automattic/wp-calypso/pull/89874, here's what I can see that has changed:

- In my PR I used `useProducts()` to find the new plan's price, and in this PR you used `usePricingMetaForGridPlans()`. I'm not sure of the difference right now but I'll look into it some more. From a quick look, it seems that `usePricingMetaForGridPlans()` queries `/plans` instead of `/products`. It should be easy to switch to that function if you believe it will be more accurate.
- In my PR I did not touch the data in Redux, focusing only on fetching the data directly, and in this PR you removed the `getPlanPrice()` selector entirely. I've confirmed that the selector is unused, so it could definitely be removed but I'll leave that for a separate PR.

**Comment by chriskmnds** [comment]
Hi @sirbrillig 

> As for the bug that @aneeshd16 mentioned above, I believe #90893 should fix it, but I'll wait for your review to be sure.

I'm not familiar with what changes are introduced in #90893 , but from my bit of exploration, it sounds like this PR would have fixed the issue with the original text if we switched from `item_subtotal_integer` to `item_original_cost_integer`. I did not get around to confirming completely though.
 
> * In my PR I used `useProducts()` to find the new plan's price, and in this PR you used `usePricingMetaForGridPlans()`. I'm not sure of the difference right now but I'll look into it some more. From a quick look, it seems that `usePricingMetaForGridPlans()` queries `/plans` instead of `/products`. It should be easy to switch to that function if you believe it will be more accurate.

`usePricingMetaForGridPlans` is the hook we've been pushing forth for replacing all of the current/Redux-based plan pricing selectors. However, as long we are not using Redux state, then it's ok. I'd leave it to your discretion.

> * In my PR I did not touch the data in Redux, focusing only on fetching the data directly, and in this PR you removed the `getPlanPrice()` selector entirely. I've confirmed that the selector is unused, so it could definitely be removed but I'll leave that for a separate PR.

Ok. Please feel free to remove it then.

**Comment by sirbrillig** [comment]
>  it sounds like this PR would have fixed the issue with the original text if we switched from item_subtotal_integer to item_original_cost_integer. I did not get around to confirming completely though.

That would be amazing. I'd love to be wrong, but I found that in my testing that was not the case because this upsell is trying to predict the cost of the product in the future, and that is hard to do reliably since different discounts behave differently. For example, sale coupons on domains apply only to the first year of the product, but coupon codes apply to all the years of a product. Introductory offers can have even more complex behavior. 

We could easily calculate the price difference between the _undiscounted_ products and tell the user that number, and maybe that's what you were suggesting. It wouldn't necessarily be accurate from the user's perspective, but it would avoid needing to guess.

> if we switched from item_subtotal_integer to item_original_cost_integer

I implemented your suggestion in this PR: https://github.com/Automattic/wp-calypso/pull/90957 and you can see that even with a one year plan, the price difference reads "Pay an extra $26" even though the difference in the final carts is actually $32.60. Maybe there's something I'm misunderstanding, though.

With https://github.com/Automattic/wp-calypso/pull/90893             |  With https://github.com/Automattic/wp-calypso/pull/90957
:-------------------------:|:-------------------------:
with-90893 |  with-90957

In https://github.com/Automattic/wp-calypso/pull/90893 I've basically improved the guessing to make it accurate for more cases. Since it doesn't require a ton of logic to guess, I think we should try to guess as best as we can to give the user the most accurate number, but I'm interested in what you think.

> usePricingMetaForGridPlans is the hook we've been pushing forth for replacing all of the current/Redux-based plan pricing selectors. However, as long we are not using Redux state, then it's ok. I'd leave it to your discretion.

I don't fully grok the difference between `/plans` and `/products` but I believe they more-or-less report the same prices (I've done [some testing here](https://github.com/Automattic/wp-calypso/pull/90957/commits/6def75253b56e30628303563e2f4d047f91801c2) but I'll look more deeply). If you don't have an opinion, I'd be inclined to stay with `useProducts()` for now until there's an explicit reason to switch.

> Ok. Please feel free to remove [`getPlanPrice`] then

Will do! https://github.com/Automattic/wp-calypso/pull/90958 should do it.

**Pull request closed without merging**
